### PR TITLE
feat(service): Added TLS and token policy helper

### PR DESCRIPTION
**Pull request by slot**
- Added TLS server that will automatically come up as soon as you set TLS_CERTFILE and TLS_KEYFILE. Both files need to be stored on the FS accessible by the service. In case files are not there or not readable the server will come up in http mode. The idea is to be secure by default so in case TLS is possible there is no http server starting. This is also the reason for not having a https port that needs to be configured. And last but not least it makes it backwards compatible. The only difference for existing missy services is that you will see a waring on the log that the service starts without TLS. But when it starts with TLS the services who call the newly secured service have to a) call the service with `https://` and b) they need to know the public cert of the certificate authority. This can be set with through `TLS_CAFILE`. The CA will be added to the system pool and can be used in the missy client wrapper around the HTTP client of the standard library.
- Added a helper function `service.TokenHasAccess(request,context,policy)` that will check if a policy is attached to the token in the request context. It should be pretty fault tolerant and solid so it will return `false` also on malformed tokens or when the token is missing completely

### CODE REVIEW

File: service/helpers.go [hunk 1]
```diff
@@ -17,3 +19,48 @@ func Vars(r *http.Request) map[string]string {
...
+	token := Token(r)
+	// return false if there is no token
+	if token == nil {
+		return false
+	}
+	// let's assume the claims are map claims because this is what our IAM delivers
+	claims, ok := token.Claims.(jwt.MapClaims)
+	// if the claims do not contain policies return false
+	if !ok {
+		log.Warn("Invalid token format: token claims are not of type jwt.MaoClaims")
```
Comment (levrik) [inline service/helpers.go]: Typo in `MapClaims`

File: service/service.go [hunk 2]
```diff
@@ -308,7 +317,7 @@ func (w *ResponseWriter) WriteHeader(code int) {
 	w.ResponseWriter.WriteHeader(code)
 }
 
-// Just sets the status code for metrics and logging
+// WriteMetricsHeader just sets the status code for metrics and logging
 // Useful for example for hijacked ResponseWriter where WriteHeader and Writer are bypassed
```
Comment (levrik) [inline service/service.go]: Typo by me, maybe you can fix it. `Writer` -> `Write`

File: service/helpers.go [hunk 3]
```diff
@@ -17,3 +19,48 @@ func Vars(r *http.Request) map[string]string {
...
+// TokenHasAccess checks if a valid access token contains a given policy in a context
+func TokenHasAccess(r *http.Request, context string, policy string) bool {
+	token := Token(r)
+	// return false if there is no token
+	if token == nil {
+		return false
+	}
+	// let's assume the claims are map claims because this is what our IAM delivers
+	claims, ok := token.Claims.(jwt.MapClaims)
+	// if the claims do not contain policies return false
```
Comment (levrik) [inline service/helpers.go]: I think this comment should be above line 37 ?

File: service/helpers.go [hunk 4]
```diff
@@ -17,3 +19,48 @@ func Vars(r *http.Request) map[string]string {
...
+	if !ok {
+		log.Warn("Invalid token format: token claims are not of type jwt.MapClaims")
+		return false
+	}
+	contextPolicies, ok := claims["policies"].(map[string]interface{})
+	if !ok {
+		log.Warn("Invalid token format: policies inside claims are not of type map[string]interface{}")
+		return false
+	}
+	// if the policies does not contain the context return false
```
Comment (levrik) [inline service/helpers.go]: Type `does` -> `do`

**Comment by pbrzostowski** [comment]
can we have `coveralis` to be more tolerant on coverage drops?

**Pull request merged**